scripts/generateData.sh

```diff
@@ -0,0 +1,289 @@
+  esac
+done
+
+# reset getopts counter again
+shift $(($OPTIND - 1))
+
+# require output folder argument and check if it is a directory
+if [ -z "${outputFolder}" ]; then
+  echo "You must specify an output directory!" >&2
+
+  printHelp
+  exit 1
+else
+  if [ ! -d "${outputFolder}" ]; then
+    echo "${outputFolder} is not a directory!"
+
+    printHelp
+    exit 1
+  fi
+fi
+
+scriptDir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+# user info about to be generated stuff
+customers=$(( n ))
+c_storeVisits=$(( k * 4 ))
+groups=$(( n / 20 + 1 )) # bash only supports integer division!
+g_discounts=$(( m / 20 ))
+carts=$(( k * n ))
+c_purchases=$(( (groups * g_discounts) / carts ))
+storeSections=$(( m / 400 + 1))
+s_inventory=$(( 400 ))
+s_purchases=$(( 2 * (carts * c_purchases) / storeSections ))
+
+echo "Generating data in CSV format for load testing:"
+echo "  ${customers} customers"
+echo "  ${m} items"
+echo "  $(( carts * c_purchases )) purchases in ${carts} carts"
+echo ""
+echo "Distributing data equaly to dactors:"
+echo "  #customers = ${customers}"
+echo "    #storeVisits = ${c_storeVisits}"
+echo "  #groups = ${groups}"
+echo "    #discounts = ${g_discounts}"
+echo "  #carts = ${carts}"
+echo "    #cart_purchases = ${c_purchases}"
+echo "  #storeSections = ${storeSections}"
+echo "    #inventory = ${s_inventory}"
+echo "    #purchase_history = ${s_purchases}"
+echo ""
+echo "Output folder: ${outputFolder}"
+echo ""
+
+function open_file() {
+  local dactor=$1
+  local id=$2
+  local relation=$3
+  declare -n file=$4
+
+  folder="${outputFolder}/${dactor}-${id}"
+  mkdir -p ${folder}
+  file="${folder}/${relation}.csv"
+}
+
+rand_counter=0
+function rand() {
+  local min=$1
+  local max=$2
+
+  # too slow
+  #echo $(python -S -c "import random; print random.randrange(${min},${max})")
+
+  # external script
+  # still too slow --> 34.8m for default arguments (see above)
+  #echo $("${scriptDir}/rand.sh" "$@")
+
+  # pseudo-rand
+  # takes only 2.1m for default arguments
+  local next=${rand_counter}
+  (( ++rand_counter ))
+  echo $(( min + (next % max) ))
+}
+
+# headers:
+# - StoreSection
+inventoryHeader="i_id,i_price,i_min_price,i_quantity,i_var_disc"
+purchaseHistoryHeader="i_id,time,i_quantity,c_id"
+
+# - GroupManager
+discountsHeader="i_id,fixed_disc"
+
+# - Customer
+customerInfoHeader="cust_name,c_g_id"
+storeVisitsHeader="store_id,time,amount,fixed_disc,var_disc"
+passwdHeader="enc_passwd"
+
+# - Cart
+cartInfoHeader="c_id,store_id,session_id"
+cartPurchasesHeader="sec_id,session_id,i_id,i_quantity,i_fixed_disc,i_min_price,i_price"
+
+
+# generate inventory data
+items=( )
+
+i=0
+while [ ${i} -lt ${m} ]; do
+  quantity=$(( i * 10 ))
+  items[${i}]="${i},${i}.${i},${i}.${i},${quantity},${i}.5"
+  (( ++i ))
+done
+#echo "number of items ${#items[@]}"
+
+# default time value
+timeValue="2017-05-02T11:01:32Z"
+
+
+# write store sections
+echo "Writing store sections ..."
+i=0
+while [ ${i} -lt ${storeSections} ]; do
```

Comment:
Can you put each loop for writing `StoreSections`, `Carts`, `Customers`, etc. in subprocesses?